Improve Variable interface

goldsborough opened the pull request:
This PR is a large change that revamps the `autograd::Variable` interface to provide a clean API that encapsulates implementation details and reduces code required to interact with `Variable`s.

A large part of this PR are small changes to update use-sites to use the new API. I would recommend beginning by taking a look at `variable.h` and `variable.cpp`, as well as `variable_version.h`, which have large changes.

I've also attempted to improve documentation of the `Variable`. Please let me know if either the API or the documentation does not make sense, or if either could be improved.

@zdevito @apaszke @colesbury

goldsborough committed:
Improve Variable interface

apaszke commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -35,11 +41,13 @@ static PyObject* THPVariable_NewWithVar(PyTypeObject* type, Variable var)
-    v->cdata.get()->pyobj = obj;
-    if (auto fn = dynamic_cast<PyFunction*>(v->cdata.get()->_grad_fn.get())) {
+    v->cdata.set_pyobj(obj);
+    if (auto fn = std::dynamic_pointer_cast<PyFunction>(v->cdata.grad_fn())) {
```
This previously didn't require a temporary incref, but now it does

colesbury replied:
I think this also will modify the `_grad_fn` on stale views, which I don't think we want to do here.

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,150 +1,255 @@
+
+  bool operator!=(const Variable& other) const noexcept {
+    return !(*this == other);
+  }
```
I don't think that's a great idea. Not sure what this overload does in ATen, but I'd expect `tensor1 == tensor2` to give me back a bool tensor of the same size, which tells me what elements are equal. Making it different for `Variables` would be very inconsistent

goldsborough replied:
Ok. It replaces code that was 'tensor1.get() == tensor1.get()'. I can make it `is_equal()`.

colesbury replied:
`is_equal` looks to much like `bool equal(const Tensor & other)` to me, which compares values. Maybe `is_same`?

goldsborough replied:
Ok

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,150 +1,255 @@
-  inline const Variable & grad() const;
-  inline       Variable & grad();
+  /// Set the type of the underlying `Tensor`.
+  void set_type(at::Type*);
```
Uuuum, what? How can you set a type of the tensor without touching the tensor? Is it a type cast or what?

goldsborough replied:
This replaces [this](https://github.com/pytorch/pytorch/blob/a8bda67ff1f4229077342c429f94673b919b9e8f/torch/csrc/autograd/python_variable.cpp#L264) rather gross hack, which should definitely go away (I didn't even know it was possible at all to set private members in C++). Is there a better way?

apaszke replied:
I agree this an awful piece of code, but making a hack part of a public API kind of makes it seem nice while what you're doing is terrible. I don't think there's a clean way to solve this, so I'd rather have the hack in the single place where it has to be, and don't imply that it's safe to do this in general.

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -159,171 +264,135 @@ struct VariableViewImpl : public VariableImpl {
-  int attr_version;
+  // grad_fn field is stale if attr_version !=
+  // version_counter.current_version()
+  uint32_t attr_version = 0;
```
this doesn't really need to be initialized, and it's not true that it's always zero when a view is created. Adding initialization makes it more confusing

goldsborough replied:
Ok, I'll remove it.

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,150 +1,255 @@
+// due to in-place modifications of the shared data. Accesses should go
+// through get_grad_fn(). All other fields are always valid.
+struct Variable::ViewImpl : public Variable::Impl {
+  ViewImpl(Variable base_, at::Tensor data_, Edge gradient_edge);
```
Can we not call arguments with an underscore at the end? We use this convention for naming struct members in certain places, so let's avoid this

goldsborough replied:
But then in the constructor it's ambiguous if you have `: base(base)`?

apaszke replied:
idk, is it? It's quite a common idiom

apaszke replied:
Alternatively how about adding them at the beginning to avoid confusion?

goldsborough replied:
I renamed them to `base` (without underscores).

goldsborough replied:
This was a bad idea. The variables shadow each other in nasty ways. I'll put the underscores in the definition and not the declaration, if that helps.

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -159,171 +264,135 @@ struct VariableViewImpl : public VariableImpl {
+}
 
-inline Variable::Variable(VariableImpl * self, bool retain) : Tensor(self, retain) {
+inline PyObject* Variable::pyobj() const noexcept {
```
Is there a reason why you only mark so few things `noexcept`? E.g. `set_pyobj` can't throw either

goldsborough replied:
I do it rather mechanically for getters since they rarely throw. I'll take another look at other methods, thanks.

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,150 +1,255 @@
+  // know are Variables.
+  /*implicit*/ Variable(at::Tensor const& rhs) : at::Tensor(rhs) {}
+  /*implicit*/ Variable(at::Tensor&& rhs) noexcept
+      : at::Tensor(std::move(rhs)) {}
```
I don't think giving up `make_variable` for an extra set of constructors is a good idea. Right now the only thing that makes constructors which take *raw tensors* and those that take *tensors assumed to be Variables* (with no checking!!) is a single `bool` argument. That's a very unsafe API. `make_variable` existed precisely for the purpose of making the distinction clear

goldsborough replied:
Ah ok. It just seemed odd to me to have to write `vector.emplace_back(make_variable(...))` instead of just `vector.emplace_back(...)`. But I get your point. `make_variable` just sounded like it should really be a constructor though. Maybe we should make them static functions and call them `Variable::wrap(...)` instead?

ezyang replied:
The `make_` convention is not without precedent in C++. For example, you don't have a constructor for both `T*` and the perfect forwarding version of `std::shared_ptr`, you have a constructor and `std::make_shared` (in this case, you really don't want an overload because they interact poorly with perfect forwarding (see Effective Modern C++ for more details ;)). I do understand static functions are common in pure-OO codebases like Java, where top-level functions are not actually a thing, but the prevailing style in PyTorch is to prefer plain old functions over static methods.

apaszke replied:
Yeah, I think sticking to the `make_` C++ convention is nicer than a static method

goldsborough replied:
Ok, I'll bring back the `make_*` functions.

goldsborough commented on torch/csrc/utils/variadic.h:
```diff
@@ -84,4 +89,23 @@ inline size_t count_variables(Args&&... args) {
+//                std::index_sequence shim for C++11
+//===----------------------------------------------------------------------===//
+
+// A container of type-template parameter indices.
```
@ezyang I remember you mentioned in one of your posts in our C++ groups that you were missing index_sequence. Idk if this helps you?

ezyang replied:
Yes, I ended up implementing tuple recursion manually for `TuplePacker`.

Curious about a few things: (1) did this code come from somewhere or was developed from first principles; (2) how does it compare to the C++14 integer sequence, is it exactly compatible but under a different name, or is there porting involved?

goldsborough replied:
I came up with this myself (I've done a fair amount of template metaprogramming in the past). `std::index_sequence` will amount to the same thing, but it's implemented far more trivially (you can [see it here](https://github.com/llvm-mirror/libcxx/blob/5b12e3bdb43ac545206501fae955b37cef0c7475/include/utility#L839)) because in C++14 you have templatized typedefs. The core idea is to get a type with a variadic list of integers so that you can grab them with `size_t... Is` in a template parameter list and then use them in a pack expansion

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,150 +1,255 @@
+/// A `Variable` augments a `Tensor` with the ability to interact in our
+/// autograd machinery. `Variable` inherits from `Tensor` and may be converted
+/// to and from `Tensor` implicitly.
+class Variable : public at::Tensor {
```
Why make it a `class` if it worked fine as `struct`, and was mentioned like this in so many places? `Function` is still a struct, so it will only be inconsistent.

goldsborough replied:
In most codebases `class` is used when `public` and `private` fields are used, to denote that this class/structure represents an abstraction, while `struct`s are used to signify a plain-old datatype, basically a pure data struct, with few methods and all fields exposed ([reference](https://stackoverflow.com/questions/54585/when-should-you-use-a-class-vs-a-struct-in-c)). I felt like using the same idiom in our codebase would be nice, and I was going to rename `Function` into a class next as I brush it up too. But I can change it to `struct`. What do you think?

apaszke replied:
I'd rather stick to our current conventions instead of changing them. Having some types as `struct`s and other as `class`es is very annoying, because you have to remember which are which when forward declaring them, friending, etc. I'd rather keep them as `struct`s to be consistent

apaszke commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -1,46 +1,68 @@
+
+  // Must be defined externally to avoid it being inlined by the compiler,
+  // which would require it to see the definition of ValueTracingState.
+  ~SavedVariable();
```
Why? It was automatically defined previously and worked fine

goldsborough replied:
I forward declared `ValueTracingStateElem` so we don't have to include `tracer_state.h` which includes the whole JIT IR. Holding a `shared_ptr`/`unique_ptr` to an incomplete type works only if the destructor is defined in the implementation file (where the definition of `ValueTracingStateElem` is available), as the inlined destructor will otherwise call the destructor of the smart pointer, which will require the size/definition of the forward declared class.

zdevito replied:
Does this mean that Variable.h will no longer include ir.h? That is a big win for compilation times.

ezyang replied:
This feels like tail wagging the dog. Wouldn't you rather have a complete definition for `ValueTracingStateElem` and incomplete types for `TracingState` and `Value`? (does the destructor problem apply to `weak_ptr` too?)

apaszke replied:
Yeah I'd rather have `ValueTracingStateElem` to be visible, with a forward declaration of required JIT types then have to change the Variable.

apaszke commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -1,46 +1,68 @@
-  Variable unpack(std::shared_ptr<Function> saved_for=nullptr) const;
-};
+  std::shared_ptr<Function> grad_fn_;
+  bool has_grad_fn_ = false;
```
Can you compress those bools in one place? Right now the struct will likely have 1B for a bool, 7B for padding, 8B for grad_fn, 1B for bool, 7B for padding, ...

goldsborough replied:
Sounds like a good idea, will implement.

apaszke commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -1,49 +1,61 @@
+
+Variable SavedVariable::unpack(std::shared_ptr<Function> saved_for) const {
+  if (!data_.defined()) {
+    if (!was_default_constructed_) {
```
What was wrong with using `version` to check this? Why did we loose `SavedVersion`? It was a nice API that encapsulated the version checking that you have to do manually now

goldsborough replied:
I began by trying to untangle the mutual `friend`-ness of `SavedVariable` and `VariableVersion`. For this I had to move the `VariableVersion` pointer out of the `SavedVariable`. At that point the struct became a bit useless. I also did some static analysis in my head to find that `saved_version.defined()` would only be false when the `SavedVariable` is default constructed. It's like 100 lines less code overall so I thought it would be nice.

ezyang replied:
I'm confused by your statement "mutual friend-ness of SavedVariable and VariableVersion". SavedVariable does not have any friends.

goldsborough replied:
It was `VariableVersion` and `SavedVersion`, sorry. The two classes that used to be in this file.

ezyang replied:
I'd love to know what the new relationships between Variable, SavedVariable, VariableVersion, et al are.

goldsborough replied:
From the viewpoint of the rest of the code the relationships are still the same. I just turned the `SavedVersion` struct into a `uint32_t`. `SavedVersion` previously had a back-reference to the `VariableVersion` so that it could:
1. check if the version it stored differs from the current version of the `VariableVersion` it references. This has been replaced by the `saved_version_ != version_counter_.current_version()` that is now in `SavedVariable` and
2. check if the `SavedVersion` is `defined()` (by checking if the `version_block_` of the `VariableVersion` it references is `nullptr`). Since `VariableVersion` always has a non-`nullptr` `version_block`, the latter would only be the case if `SavedVersion` was default-constructed. In turn, it would only be default-constructed if the `SavedVariable` was default constructed. So I replaced this whole behavior with the `was_default_constructed` boolean.
This was an implementation detail of `SavedVariable` and `VariableVersion`, so the rest of the code will still construct `SavedVariable`s from `Variable`s as before.

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -1,102 +1,140 @@
+      data(std::move(data_)),
+      grad_fn(std::move(gradient_edge.function)),
+      requires_grad(requires_grad_),
+      output_nr(gradient_edge.input_nr) {
```
I don't feel like having initialization of some fields specified in the header, while others mentioned here is a good idea. It makes it very easy to forget to initialize a certain fields, especially that those structs are very complicated.

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -106,52 +144,84 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, int output_
+
+void Variable::rebase_history(
+    int output_nr,
+    std::shared_ptr<Function> grad_fn) {
```
Everything accepts `Edges` now, except this function. Would be good to make it consistent

goldsborough replied:
Thanks

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -106,52 +144,84 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, int output_
+    impl.rebase_history(output_nr, std::move(grad_fn));
+  } else {
+    get()->output_nr = output_nr;
+    get()->grad_fn = std::move(grad_fn);
```
This can use `set_gradient_edge`

goldsborough replied:
Thanks!

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -106,52 +144,84 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, int output_
+
+bool Variable::requires_grad() const noexcept {
+  return get()->requires_grad || get()->grad_fn ||
+      (is_view() && base().requires_grad());
```
Why is this no longer in the header? It's used all over the place, and it should be inlined.

goldsborough replied:
My bad, will put back.

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -106,52 +144,84 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, int output_
+  if (is_view()) {
+    return static_cast<Variable::ViewImpl*>(get())->base;
+  }
+  throw std::runtime_error("Can't get base of non-view");
```
This could also be in the header

apaszke reviewed (changes requested):
I'm not a huge heavyweight OOP fan, so I'm a bit ambivalent about this PR. Some changes just pack the same code in slightly nicer methods, while I'd consider other to be quite unnecessary (e.g. `struct` -> `class` that doesn't really fit the conventions of our codebase) or dangerous (removing `make_variable`). I'd also like to have @colesbury look over them to make sure I haven't missed anything, since this is a critical part of this system.

colesbury commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,150 +1,255 @@
-  inline VariableImpl* get() const;
+  /// Return the input index of the gradient `Function` to which this `Variable`
+  /// is connected.
+  int output_nr() const noexcept;
```
The advantage of marking these declarations inline in the header is that you get compilation errors (instead of runtime or linker errors) if you miss the definition below.

goldsborough replied:
Ah interesting, didn't know that. I felt like it was a bit of "visual noise" in the declarations. Do you think we should put them back?

colesbury commented on torch/csrc/autograd/functions/special.cpp:
```diff
@@ -264,11 +265,10 @@ bool Eval::replaceSubgraph(const variable_list& inputs, const variable_list& _ou
     // the same Variable has been returned multiple times, and
     // is repeated in this list.
-    if (output.get()->_grad_fn.get() == this) {
+    if (output.grad_fn().get() == this) {
```
This change may have performance implications: `grad_fn()` will update the _grad_fn if output is a view with a stale _grad_fn. I don't think that's what you want for this comparison.

goldsborough replied:
Thanks, I've created a `grad_fn_ptr()` class which returns the raw pointer.

apaszke replied:
How about `grad_fn_unsafe()`? It's unclear what's the difference between `grad_fn` and `grad_fn_ptr`, while it really is unsafe to use the second method without knowing what you're doing

colesbury commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -106,52 +144,84 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, int output_
+}
+
+Variable::Impl* Variable::get() const {
+  return static_cast<Variable::Impl*>(pImpl);
```
Can this be in the header?

colesbury reviewed (commented):
This looks generally reasonable to me. I'm most concerned about:

1) The Variable() constructor which takes data. I think a static factory method might be better
2) Movement of some functions to the variable.cpp. We don't compile with LTO and many of these are called often.

goldsborough commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -1,46 +1,68 @@
+#include <ATen/Tensor.h>
+
+#include <cstdint>
+#include <list>
```
Ten lines below this

colesbury replied:
What about the other files? Like `torch/csrc/utils/tuple_parser.cpp`

goldsborough commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -1,46 +1,68 @@
-#include "torch/csrc/Types.h"
 
-namespace torch { namespace autograd {
+#include <ATen/Tensor.h>
```
Ok 👍

goldsborough committed:
Address comments from @apaszke and @colesbury

goldsborough commented:
I think I got all your comments. Let me know about `make_variable` vs. `Variable::<something>`.

goldsborough committed:
string ::operator= is not noexcept

ezyang commented on tools/autograd/templates/VariableType.cpp:
```diff
@@ -378,9 +362,8 @@ template<typename... Args> inline variable_list flatten(Args&&... args) {
-  auto& var = static_cast<const Variable&>(t);
-  var.version_counter().increment();
+static void increment_version(Tensor & t) {
+  static_cast<Variable&>(t).bump_version();
```
While we have the patient open, can we centralize all occurrences of `static_cast<Variable&>(tensor)` into an inline function `Variable& castToVariableRef(Tensor&)` so that we can add a test that Tensor really is a Variable in DEBUG mode?

ezyang commented on tools/autograd/templates/VariableType.cpp:
```diff
@@ -207,49 +207,35 @@ static std::vector<SavedVariable> make_saved_variable_list(TensorList tensors) {
+  // sizeof...(Tensors) - 1.
+  return as_variable_impl(
+      tensors, typename MakeIndices<sizeof...(Tensors)>::indices());
 }
```
TuplePacker in `tools/jit/templates/aten_dispatch.cpp` will also benefit from indices machinery.

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -282,7 +292,7 @@ int THPVariable_set_data(THPVariable *self, PyObject *data)
   if (&self->cdata.data().type() != &tensor.type()) {
     // we change the type of var.data so we must change the type of var
     auto newType = VariableType::getType(tensor);
-    self->cdata.get()->*get(TensorImpl_Type()) = newType;
```
There are no more call sites of `TensorImpl_Type`. This means you should delete the hack, right?

```
// XXX: This is a hack to access private TensorImpl::type_
// http://bloglitb.blogspot.com/2011/12/access-to-private-members-safer.html
// This is currently needed because module.float() changes the type of the
// data field of each variable. We should fix this and not allow changing the
// type of var.data.
```

But this is not the end of the story. See `set_type` for more.

zdevito commented on torch/csrc/autograd/functions/basic_ops.cpp:
```diff
@@ -5,6 +5,8 @@
 #include "torch/csrc/autograd/functions/utils.h"
 #include "torch/csrc/utils/auto_gpu.h"
 
+#include <ATen/Tensor.h>
```
Only `#include <ATen/ATen.h>` is actually correct, for not-so-good but not-easy-to-fix reasons. We may eventually move all the other headers so  ATen/ATen.h is the only thing that works.

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,150 +1,250 @@
-  inline const Variable & grad() const;
-  inline       Variable & grad();
+  /// Set the type of the underlying `Tensor`.
+  void set_type(at::Type*) noexcept;
```
I noticed this in code review because previously there was no setter for type, and now there is. That should have indicated that something is up.

In fact, the back story is that `set_type` is exposed *solely* to support type-changing of Variable (documented at:

```
// XXX: This is a hack to access private TensorImpl::type_
// http://bloglitb.blogspot.com/2011/12/access-to-private-members-safer.html
// This is currently needed because module.float() changes the type of the
// data field of each variable. We should fix this and not allow changing the
// type of var.data.
```

So, there should (1) be a very clear indication that `set_type` is really exposed for OO reasons, but not as a legitimate thing you should actually do with a Tensor (unless we revisit the design); at least prefixing it with an underscore but also a comment and (2) a clear reference to the portion of code which actually makes use of it, and the reason why (namely, to support parameter identity-preserving, inplace type casting of modules). You can use the `Note [Foo]` convention to do this.

ezyang commented on aten/src/ATen/TensorImpl.h:
```diff
@@ -16,6 +16,10 @@ struct Storage;
   explicit TensorImpl(Type * type)
   : is_scalar(false), type_(type) {}
+
+  void setType(Type* type) noexcept {
```
Reference to `Note [...]` here too.

zdevito reviewed (commented):
I have a few comments, not yet a full review.

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -153,177 +253,162 @@ struct VariableViewImpl : public VariableImpl {
-  return get()->version_counter;
+inline jit::tracer::ValueTracingState& Variable::tracing_state() const
+    noexcept {
+  return *get()->tracing_state;
```
This function has a precondition now, that `get()->tracing_state` is not null (and it certainly may be null, for untraced variables). The precondition should be documented, and if you don't want a user to see a segfault if we get it wrong, you should add an assert to this effect too.

goldsborough replied:
Actually the `tracing_state` is an `auto_unique_ptr` so it will get constructed on dereference

ezyang commented on torch/csrc/jit/graph_executor.cpp:
```diff
@@ -133,15 +134,9 @@ struct ExecutionPlan {
-      // Note: we have to set this up in place, or we have to
-      // throw away and reallocate variables that were already created in
-      // wrapTensors. We should add an API for this, and more generally
-      // we need to clean up the fields of Variable.
```
Isn't the first half of this comment still applicable?

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -35,11 +41,13 @@ static PyObject* THPVariable_NewWithVar(PyTypeObject* type, Variable var)
-      v->cdata.get()->_grad_fn = THPFunction_asFunction((THPFunction*)fn->obj);
+      const auto output_nr = static_cast<uint32_t>(v->cdata.output_nr());
+      v->cdata.set_gradient_edge(
+          {THPFunction_asFunction((THPFunction*)fn->obj), output_nr});
```
This looks semantically different, what's going on here?

goldsborough replied:
Yeah this one was a bit confusing. Previously we were only setting `grad_fn` and leaving the `output_nr` as it was. To be able to get rid of `set_grad_fn` in favor of `set_gradient_edge()`, I figured the equivalent would be to get the `output_nr` from before and set it as the new `output_nr` in the gradient_edge. So I believe the net-effect is the same? And I think `set_gradient_edge()` is the better API everywhere else since it seems less reasonable to change the underlying `grad_fn` without changing the `output_nr` -- we'd rather use the concept of an `Edge`.

goldsborough committed:
Remove ir.h from tracer_state.h to improve build times

goldsborough committed:
Make Variable a struct and pack SavedVariable fields

goldsborough committed:
Implement as_variable_ref

goldsborough committed:
grad_fn_ptr() -> grad_fn_unsafe()

goldsborough committed:
Reduce hackiness of set_type hack

goldsborough commented:
Summary of updates and who this should be interesting for:
- Made `Variable` constructors into free functions @apaszke  @ezyang 
- Implemented `as_variable_ref` to downcast `Tensor` to `Variable`, with a safe mode if `DEBUG` is defined @ezyang 
- Changed `grad_fn_ptr()` to `grad_fn_unsafe()` @apaszke 
- Made the `set_type` hack less hacky by (1) Making the `type_` field protected in `TensorImpl` (@zdevito says this is fine) and makes sense anyway and (2) prefixing `set_type` with `temporary_hack_`, @ezyang @apaszke 
- Slightly unrelated to other things: removed the `ir.h` include from `tracer_state.h`, so that future modifications to the JIT will not rebuild `variable` and its descendants (70 files or so) @zdevito

goldsborough committed:
Include variable.h and edge.h in tracer_state.h because it uses them

goldsborough committed:
class Variable -> struct Variable because Windows cant even

goldsborough committed:
Make Variable::output_nr uint32_t instead of int

goldsborough committed:
Add comment about tracing state

goldsborough committed:
Replaced more static_cast<Variable&> and improve docs

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,329 +1,615 @@
+/// `Function`s in the autograd graph. A `Variable` can either be a leaf, like a
+/// weight in a neural network, or an interior variable, when it is the result
+/// of an operation between variables. Every `Variable` also stores another
+/// `Variable` (recursively) called its `grad` (gradient). If the variable is a
```
it's rarely recursive, since `.grad`s are usually not leaves, and so they don't get gradient saved by default (`.grad.grad` is not a very useful value - it's not the same as taking the second-order derivatives)

goldsborough replied:
Ok, thanks for clarifying. I'll remove the word

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,329 +1,615 @@
+/// `Variable` and `Tensor` actually convert implicitly between each other. You
+/// can thus call functions defined on `Tensor`s also with `Variable`s. Besides
+/// the constructor of `Variable` that converts to it from `Tensor`, you can use
+/// the `make_variable` free functions to create variables. To create views,
```
This implies that you can just use the "constructor that converts a Tensor into a Variable or `make_variable`", but they are not really equivalent - one is unsafe and *expects an upcasted Variable*, the other one expects a *raw tensor* and will wrap it in a Variable

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,329 +1,615 @@
+  /// i.e. a (function, input_nr) pair specifying the function in the autograd
+  /// graph, and what particular input of that function, this variable is
+  /// connected to.
+  friend Variable make_variable(at::Tensor data, Edge gradient_edge);
```
Can you please mention that `data` *should not be a Variable* in the comment?

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,329 +1,615 @@
+
+  /// Update the grad_fn of an existing Variable. Called after in-place
+  /// modifications.
+  void rebase_history(Edge gradient_edge);
```
Can you mention this function next to `set_gradient_edge`? It's important that this one is used when in-place modifications are involved

apaszke commented on torch/csrc/autograd/variable.h:
```diff
@@ -1,329 +1,615 @@
+#ifdef DEBUG
+  // dynamic_cast will return a nullptr if the `TensorImpl`'s dynamic type is
+  // not `Variable::Impl`.
+  if (dynamic_cast<Variable::Impl*>(tensor.get()) == nullptr) {
```
This can be a `ViewImpl` too right?

goldsborough replied:
Should be fine I believe because `ViewImpl` inherits from `Impl` so `dynamic_cast`ing a `ViewImpl` to `Impl` would still work (i.e. return non-`nullptr`)

apaszke replied:
Ah I wasn't aware of the inheritance. Never mind the comment then

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -106,52 +105,71 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, int output_
-  Variable detached = make_variable(data());
-  detached.version_counter() = version_counter();
+  auto detached = make_variable(data(), /*requires_grad=*/false);
+  detached.set_version(version_counter());
```
(irrelevant to this PR) @colesbury do you remember why do we detach the version counter as well? I don't think this should ever happen, since that will break correctness checks

colesbury replied:
We share the version counter when we create a detached Variable since they share the same data. That way, if you modify the detached variable, you'll still get the error message in backwards if it messes up a SavedVariable.

Is that what you mean by "why do we detach the version counter as well"?

apaszke replied:
Ugh I read `version_counter()` as a call to the constructor, not `this->version_counter()`. Sorry...

goldsborough commented on tools/autograd/templates/VariableType.h:
```diff
@@ -56,7 +60,6 @@ struct VariableType final : public at::Type {
   static at::Tensor unpack_opt(const Tensor & t, const char * name, int pos);
   static std::vector<at::Tensor> unpack(at::TensorList tl, const char *name, int pos);
 
-private:
```
The second `private` here seemed redundant because of the `private` above (everything below the `private` is private)

apaszke replied:
Sure, didn't expand the snippet, so I couldn't see the other `private`

goldsborough commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -1,49 +1,61 @@
+  version_counter_ = variable.version_counter();
+  saved_version_ = version_counter_.current_version();
+  has_grad_fn_ = !variable.is_leaf();
+  output_nr_ = variable.output_nr();
```
True. It's a bit sketchy because `defined()` might be `false`, but these values should still be well-defined nevertheless, so I'll put them there.

goldsborough committed:
Remove SavedVariable destructor and construct members in init list

goldsborough committed:
Clarify docs for Variable

goldsborough committed:
Variable::set_version -> set_version_counter

goldsborough commented:
Achievement unlocked: make clang segfault

ezyang reviewed (approved):
Let's merge this before it bitrots further.

@goldsborough I think we need another rebase

goldsborough commented:
🙏
